### PR TITLE
Features/add coherence measure

### PR DESCRIPTION
**Pull request by mrufsvold**
Add a coherence measure function to `modelutils.jl` that implements a modified UMass Coherence Score based on the findings "Exploring the Space of Topic Coherence Measures" (Röder, Both, & Hinneburg, 2015). 

At a high-level, the function takes the following steps:
1) Get keys for topic words
2) Make one2previous token pairs for each topic (for each token, create tuples with all other tokens in the topic with a higher rated salience)
3) For each pair, calculate confirmation measure: (P(prev|current) - P(prev|!current)) / (P(prev|current) + P(prev|!current))
4) Take the mean of these confirmation measures.

Röder, Both, & Hinneburg find even better performance on corpora with large documents using indirect confirmation measures (comparing the relationship between each word in the pair and all other words in the vocab). However, for my application, I have small docs so it wasn't worth any small gains I might get. I would consider taking this on later.

They also suggest using a sliding window (considering word 1 through 100 in the doc and then moving the window along the doc one word at a time) for as virtual docs. However, since this repo is already setup to use bag of words, I didn't want to introduce additional complexity in handling word order accurately. 

For what it's worth -- I ran the models you created in the tutorial through the coherence measure using the training corpus (I didn't expand the corpus to a more general set of documents) and got a coherence of 0.330 and 0.315 for the LDA and fCTM respectively. I find it surprising that fCTM didn't come up with higher coherence than LDA but using the training corpus not a more general reference for coherence could be the source of the problem.

### CODE REVIEW

**Comment by mrufsvold** [comment]
I got some feedback from a colleague on my code and refactored it in light of his suggestions. Basically, changed the make_word_pairs() function from creating a large array to returning a channel that returns one tuple at a time. Significantly reduced allocations. 

I also added a parallel version of findcoherence() but it depends on having Julia 1.7 because it uses @atomic. I have a question posted on the Discourse page to see if anyone has a more backwards compatible version of this. I'll update it if I get a suggestion. 

I also found one logical mistake I made and addressed that. I need to rerun the scores for the Readme, but I won't have time for that until next week.

**Comment by mrufsvold** [comment]
Okay, I spent some time to better understand channels and I created a threadsafe version of accumulation without using @atomic so it should be backwards compatible.

**Comment by ericproffitt** [comment]
I made some minor changes, mostly syntactic, however `buffer` is not defined in the `findcoherence` function, so this needs to be fixed before we can proceed.

**Comment by mrufsvold** [comment]
Sorry for the delay. Thanks for the catch!

**Comment by mrufsvold** [comment]
Sorry for messing up your formatting! I thought I had done that, and was trying to make it more standard-ish.

**Comment by ericproffitt** [comment]
Going forward, I'm trying to enforce a comment syntax of `## comment...`, so I reverted this part of your commit back to what I had before.

Also, there are some edge case bugs which occur when the corpus vocabulary is very small, or when it contains terms which do not exist in any of the documents, these will also need to be addressed prior to merging the pull request. The cause of this appears to be division by zero in the `calculate_confirmation` function, and at the bottom of the `findcoherence` function, which result in the output of NaNs from the `findcoherence` function.

For example, `model = LDA(Corpus(), 10)` is a model with an empty corpus, and produces a NaN when one runs `findcoherence(model)`. Similarly, both models

```julia
corp = readcorp(:citeu); corp.docs = corp[1:3]; fixcorp!(corp); model=LDA(corp, 10)
```

```julia
model=LDA(Corpus(Document(1:2), vocab=["term1", "term2"]), 10)
```

produce NaN values. In the empty case, I think the expected output of `findcoherence` should probably be 0, however in the latter cases (and other cases similar to them) I'm not sure what the appropriate functionality is.

File: src/modelutils.jl [hunk 1]
```diff
@@ -1026,13 +1037,25 @@ function one2prev_generator(topic_words::Matrix{Int}, vocab::Dict{Int, String},
     Channel{Tuple{Int, Int}}(buffer) do ch
 		for i in CartesianIndices(topic_words)
 			row, col = i[1], i[2]
+			current_tok = topic_words[i]
+
+			## Skip this topic word if it is not found in the vocabulary
```
Comment (mrufsvold) [inline src/modelutils.jl:1042]: This check and the one on line 1053 make sure to skip if the topic words are missing from vocabulary. If everything is missing from vocabulary, then no pairs are generated and line 1026 throws an error.

File: src/modelutils.jl [hunk 2]
```diff
@@ -981,3 +981,135 @@ function topicdist(model::TopicModel, doc_indices::Vector{<:Integer})
...
+            end
+        else 
+            num_wo_current += 1
+            if word_pair[2] in terms
+                num_w_prev_not_current += 1
+            end
+        end
+    end
+    
+	## If either word is completely missing from the corpus, then we skip this confirmation
```
Comment (mrufsvold) [inline src/modelutils.jl:1101]: Return a Nan if either word is totally missing from the corpus so that we can skip this pair when we accumulate at line 1016.  If they're all missing (maybe because the corpus is empty), then we raise another error at line 1027.

File: src/modelutils.jl [hunk 3]
```diff
@@ -981,3 +981,135 @@ function topicdist(model::TopicModel, doc_indices::Vector{<:Integer})
...
+		if isnan(r)
+			continue
+		else
+			sum_coherence += r
+			coherence_val_count += 1
+		end
+    end
+
+	(num_pairs > 0) || throw(error("None of the topic words appear in the vocabulary"))
+	(coherence_val_count > 0) || throw(error("None of the topic words appear in the coprus"))
```
Comment (mrufsvold) [inline src/modelutils.jl:1027]: I think that if all of the topic words are missing from the vocab or from the corpus, then raising an error is better than returning a zero because it helps diagnose what is happening more clearly.

**Review by mrufsvold — COMMENTED** [review]
I think the code comments are also sufficiently explanatory, but I wanted to address your edge cases specifically with some review comments. 

I don't have my environment set up to import my dev version of the package to test the edge cases right now, and my family is all down for the count with covid at the moment so this is as far as I can get it right now. 

I wanted to share what I have in case you prefer to keep it moving. Otherwise, I can come back to test this later and let you know when I have confirmed that I handled the edge cases you highlighted.

File: src/modelutils.jl (hunk 2 above)
Comment (mrufsvold) [reply src/modelutils.jl:1101]: I think Nan is better than zero because just because a word is missing from the corpus doesn't mean it's unrelated to the the other in its pair. Could just be a rare word that didn't make it into the corpus.

File: src/modelutils.jl (hunk 1 above)
Comment (ericproffitt) [reply src/modelutils.jl:1042]: Sure, no rush, you can revisit when you're able to.
